### DISCUSSION TITLE
Migration Pathway on OrchardCore.ContentFields Indexing

### DISCUSSION THREAD

**Discussion by jeffolmstead** [OrchardCMS/OrchardCore#9119] [Q&A]
@deanmarcussen I am wondering if you might know of a migration pathway as I haven't located one without directly writing SQL.  The majority of my sites have indexing tied to a release prior to RC2 and have index names [using this release](https://github.com/OrchardCMS/OrchardCore/blob/rc2-13929/src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/Migrations.cs) (or something a little later - point being I have index names of 'IDX_TimeFieldIndex_Time' in my SQL database. 

Then there was this [release into dev](https://github.com/OrchardCMS/OrchardCore/blob/ab76687ea4cc3dea68796bc224ac292f17656e60/src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/Migrations.cs) that changed the index naming convention over to 'IDX_TimeFieldIndex_DocumentId_Time' and has no reference to the earlier index (i.e. it didn't migrate the current indexes). That alone isn't a big deal, except it leaves orphan indexes in the system and I don't have the newer naming convention - this is what I have instead:
![image](https://user-images.githubusercontent.com/1848585/114266524-c43e7780-99c4-11eb-9adf-e13bfc9caabb.png)

Then this [migration update](https://github.com/OrchardCMS/OrchardCore/blob/9576ce94cfdb152303efcb778eba7296c2638031/src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/SQL/Migrations.cs) came along to fix an issue with the 'Time' column. It can't succeed in my database though because I have a legacy (orphan) index of 'IDX_TimeFieldIndex_Time' that is still dependent on 'Time' and thus the 'Time' field cannot be dropped and recreated. 

If there is no official plan to have an in between migration, then I can just prepare a script. But if it looks like others are going to face this, maybe there is a way to go back and migrate the old index names over to the new index names to prevent future issues. As shown in the image above, the 'Time' index is just one of may on this ContentFields index migration where it seems the naming conventions have been updated and it seems there were more indexes back in the day.

**Comment by deanmarcussen** [comment]
@jeffolmstead I would current suggest approaching this with a script for the moment, and reporting back.

I think others will also face this, but there are some challenges in preparing an official migration. 

There's been quite a history with the indexes for the sql fields feature, from problems with mysql key length limits / sql key length limits.

The issue is / has been, that we couldn't easily drop indexes because of differences between SQL 2016 + syntax and earlier versions, so including a drop index, caused people issues when they were running old versions of sql server.
What happens this, is the same as what is happening for you, they get stuck on a migration, that can't complete.

We decided at the time of altering the indexes, that the best effort we could do, was to add additional indexes, that were actually useful (the original indexes were never used, due to the way they were prepared. so entirely useless).

When I tested this two weeks ago, the Time Field issue had not presented, and I commented on the pr, that we needed to do it, but it would cause problems, as you are reporting.

See here for the last reference database https://github.com/OrchardCMS/OrchardCore/issues/8753 that I tested with (pre Time Field).

What I don't have is a sql database from pre RC2 to even experiment with.

Perhaps one solution to the issue could be a custom migration feature for the sql fields.

Maybe as you have a database that presents the issues you might consider trying to put something together?

Or the script you use may also be a very useful helper for anyone experiencing the issue.

Open to other ideas

**Reply by Skrypt** [reply]
I think this issue should be fixed by adding a SQLParser in YesSQL simply. And the index length should be fixed by using some index name generator method that would comply to the current DB restrictions. Here, when we migrate an index we would just require to drop them all and recreate them simply.

**Comment by jtkech** [comment]
> Here, when we migrate an index we would just require to drop them all and recreate them simply

Not saying that it was good to support old sql server versions, but yes, at some point we didn't drop old indexes as it would fail with old sql server versions, so leaving some orphans, and yes that here causes an issue because of the column type change. Maybe good to say in the doc to manually delete them.

But recently we agreed more or less to not support old sql versions, and we did recently some migrations that drop indexes, so no luck that we change the column type of this orphan index ;)

**Reply by Skrypt** [reply]
We agreed that with the current implementation we don't support old SQL Server versions but nothing prevents from adding support over time. By the mean time the proper resolution is to use a supported SQL Server version. I can't work on 3 PR's at the same time 😉

**Comment by deanmarcussen** [comment]
For a bit more context (because when we are discussing this we are talking about different issues, which all collide, into one big one)

The actual length of a value that can be stored in an index. This is the biggest thing that discourages, writing a dialect for pre SQL 2016, because the migrations would have to be different between dialects. And then what is developed on one database dialect might not be deployable to another database dialect. We could work around this with configurations, but it is the difficult area in supporting previous versions (for me) 
https://github.com/OrchardCMS/OrchardCore/issues/5196

The same issue, but for MySql, and the note that this is a key length issue, not the name of the index itself.
https://github.com/OrchardCMS/OrchardCore/issues/4753

The pr to YesSql resolving the length of the names of index tables (not related to key length)
https://github.com/sebastienros/yessql/pull/362

What happens for users when we use DropIndex
https://github.com/OrchardCMS/OrchardCore/issues/6210

The suggestion for docs on minimum databases (which I fully support, and think is a great idea, at least until someone decides that there is a way to handle these old versions, and writes some code)
https://github.com/OrchardCMS/OrchardCore/issues/8395

None of these help with @jeffolmstead particular issue.

The only thing I have there, is either a custom sql script, or we could potentially write a feature that is just a migration, which would attempt (with a lot of try catches perhaps) to cleanse the content fields database schema. Noting that in the case of the time field, the entire dataset of time fields would also have to be resaved.

**Comment by jeffolmstead** [comment]
To add to the discussion, I will add one more example of a migration that didn't work "on it's own". OrchardCore.OpenId migration throws this error:

2021-04-11 06:25:00.2809|PVT||6c8681ef-4d8849a435cb1f46.||OrchardCore.Data.Migration.DataMigrationManager|ERROR|Error while running migration version 7 for 'OrchardCore.OpenId'. Microsoft.Data.SqlClient.SqlException (0x80131904): The DELETE statement conflicted with the REFERENCE constraint "OrchardCorePVT_FK_OpenIdAppByRoleNameIndex_Document_DocumentId". The conflict occurred in database "orchardcore_pvt", table "dbo.OrchardCorePVT_OpenIdAppByRoleNameIndex_Document", column 'DocumentId'.

It is an interesting error as the table itself is being deleted.  In this particular case I can't delete the Foreign Key as that causes issues, likely in the 'DropReduceIndexTable' method, so I need to add it back as a dummy constraint. 

Here are the SQL scripts I had to run to get this one to go (note that 'OrchardCorePVT' is tenant specific and would need replaced)

```
-- NOTE AFTER RUNNING SCRIPTS, 'RELOAD' TENANT TO GET MIGRATION TO RUN AGAIN
-- DROP THE EXISTING FOREIGN KEY
ALTER TABLE [dbo].[OrchardCorePVT_OpenIdAppByRoleNameIndex_Document] DROP CONSTRAINT 
	[OrchardCorePVT_FK_OpenIdAppByRoleNameIndex_Document_DocumentId]
GO

-- ADD THE SAME 'KEY' NAME BACK IN AS A DUMMY CONSTRAINT
ALTER TABLE [dbo].[OrchardCorePVT_OpenIdAppByRoleNameIndex_Document] ADD CONSTRAINT
	[OrchardCorePVT_FK_OpenIdAppByRoleNameIndex_Document_DocumentId] CHECK (DocumentId > 0)
GO


-- DROP THE EXISTING FOREIGN KEY
ALTER TABLE [dbo].[OrchardCorePVT_OpenIdAppByRoleNameIndex_Document] DROP CONSTRAINT 
	[OrchardCorePVT_FK_OpenIdAppByRoleNameIndex_Document_Id]
GO

-- ADD THE SAME 'KEY' NAME BACK IN AS A DUMMY CONSTRAINT
ALTER TABLE [dbo].[OrchardCorePVT_OpenIdAppByRoleNameIndex_Document] ADD CONSTRAINT
	[OrchardCorePVT_FK_OpenIdAppByRoleNameIndex_Document_Id] CHECK (DocumentId > 0)
GO


-- DROP THE EXISTING FOREIGN KEY
ALTER TABLE [dbo].[OrchardCorePVT_OpenIdApplicationIndex] DROP CONSTRAINT 
	[OrchardCorePVT_FK_OpenIdApplicationIndex]
GO

-- ADD THE SAME 'KEY' NAME BACK IN AS A DUMMY CONSTRAINT
ALTER TABLE [dbo].[OrchardCorePVT_OpenIdApplicationIndex] ADD CONSTRAINT
	[OrchardCorePVT_FK_OpenIdApplicationIndex] CHECK (DocumentId > 0)
GO
```
